src/org/zaproxy/zap/extension/jxbrowser/ZapTabbedPanel.java

```diff
@@ -0,0 +1,99 @@
+ * 
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ * 
+ * Copyright 2017 The ZAP Development Team
+ *  
+ * Licensed under the Apache License, Version 2.0 (the "License"); 
+ * you may not use this file except in compliance with the License. 
+ * You may obtain a copy of the License at 
+ * 
+ *   http://www.apache.org/licenses/LICENSE-2.0 
+ *   
+ * Unless required by applicable law or agreed to in writing, software 
+ * distributed under the License is distributed on an "AS IS" BASIS, 
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
+ * See the License for the specific language governing permissions and 
+ * limitations under the License. 
+ */
+package org.zaproxy.zap.extension.jxbrowser;
+
+import java.awt.Component;
+
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.JTabbedPane;
+import javax.swing.event.ChangeEvent;
+import javax.swing.event.ChangeListener;
+
+import org.zaproxy.zap.utils.DisplayUtils;
+
+
+public class ZapTabbedPanel extends JTabbedPane {
+
+    private static final long serialVersionUID = 1L;
+
+    public ZapTabbedPanel() {
+        
+        this.addChangeListener(new ChangeListener() {
+            @Override
+            public void stateChanged(ChangeEvent e) {
+                setCloseButtonStates();
+            }
+        });
+
+    }
+    
+    public void addTab(String title, Icon icon, final Component c, boolean hideable, boolean visiXbleCanDelete, int index) {
+        if (index == -1 || index > this.getTabCount()) {
+            index = this.getTabCount();
+        }
+        if (icon instanceof ImageIcon) {
+            icon = DisplayUtils.getScaledIcon((ImageIcon)icon);
+        }
+
+        super.insertTab(title, icon, c, c.getName(), index);
+
+        int pos = this.indexOfComponent(c);
+        // Now assign the component for the tab
+        
+        this.setTabComponentAt(pos, new TabbedPanelTab(this, title, icon, c, hideable));
+
+        if ((index == 0 || getTabCount() == 1) && indexOfComponent(c) != -1) {
+            // Its now the first one, give it focus
+            setSelectedComponent(c);
+        }
+    }
+    
+    @Override
+    public void remove(Component c) {
+        int pos = this.indexOfComponent(c);
+        if (pos == getTabCount() - 2) {
+            setSelectedIndex(getTabCount() - 3);
+        }
+        super.remove(c);
+        setCloseButtonStates();
+    }
+
+    private void setCloseButtonStates() {
+        // Hide all 'close' buttons except for the selected tab
+        if (this.getTabCount() <= 2) {
+            // just one tab and maybe the Plus one, so dont allow the first one to be closed
+            Component tabCom = this.getTabComponentAt(0);
+            if (tabCom != null && tabCom instanceof TabbedPanelTab) {
+                TabbedPanelTab jp = (TabbedPanelTab) tabCom;
+                jp.setEnabled(false);
+            }
+            return;
```

Comment:
Instead of returning here would it be better to put the loop below in an else block?